apollo-router/src/query_planner/subgraph_context.rs

```diff
@@ -0,0 +1,361 @@
+                                            }
+                                        }),
+                                    );
+                                }
+                                return Some((item.rename_key_to.to_string(), new_value));
+                            }
+                        }
+                        None
+                    }
+                    DataRewrite::ValueSetter(_) => None,
+                }
+            })
+            .collect();
+        self.named_args.push(hash_map);
+    }
+
+    // Once all a value has been extracted for every variable, go ahead and add all 
+    // variables to the variables map. Additionally, return a ContextualArguments structure if 
+    // values of variables are entity dependent
+    pub(crate) fn add_variables_and_get_args(
+        &self,
+        variables: &mut Map<ByteString, Value>,
+    ) -> Option<ContextualArguments> {
+        let (extended_vars, contextual_args) = if let Some(first_map) = self.named_args.first() {
+            if self.named_args.iter().all(|map| map == first_map) {
+                (
+                    first_map
+                        .iter()
+                        .map(|(k, v)| (k.as_str().into(), v.clone()))
+                        .collect(),
+                    None,
+                )
+            } else {
+                let mut hash_map: HashMap<String, Value> = HashMap::new();
+                let arg_names: HashSet<_> = first_map.keys().cloned().collect();
+                for (index, item) in self.named_args.iter().enumerate() {
+                    // append _<index> to each of the arguments and push all the values into hash_map
+                    hash_map.extend(item.iter().map(|(k, v)| {
+                        let mut new_named_param = k.clone();
+                        new_named_param.push_str(&format!("_{}", index));
+                        (new_named_param, v.clone())
+                    }));
+                }
+                (
+                    hash_map,
+                    Some(ContextualArguments {
+                        arguments: arg_names,
+                        count: self.named_args.len(),
+                    }),
+                )
+            }
+        } else {
+            (HashMap::new(), None)
+        };
+
+        variables.extend(
+            extended_vars
+                .iter()
+                .map(|(key, value)| (key.as_str().into(), value.clone())),
+        );
+
+        contextual_args
+    }
+}
+
+// Take the existing subgraph operation and rewrite it to use aliasing. This will occur in the case
+// where we are collecting entites and different entities may have different variables passed to the resolver.
+pub(crate) fn build_operation_with_aliasing(
+    subgraph_operation: &SubgraphOperation,
+    contextual_arguments: &Option<ContextualArguments>,
+    schema: &Schema,
+) -> Result<Operation, ContextBatchingError> {
```

Comment:
this should _not_ return a new operation. This should be modifying an `ExecutableDocument` and returning a `Valid<ExecutableDocument>>`. That way we can be sure that the operation is semantically valid before sending the operation to subgraphs. 

This will also make serialization easier - `ExecutableDocument::serialized()` returns a formatted string.